gtsfm/utils/geometry_comparisons.py

```diff
@@ -30,138 +30,234 @@ def align_rotations(input_list: List[Rot3], ref_list: List[Rot3]) -> List[Rot3]:
+        input_list: input point3s which need to be aligned, suppose w1ti in world-1 frame for all frames i.
+        ref_list: reference point3s which are target for alignment, suppose w2ti_ in world-2 frame for all frames i.
 
     Returns:
-        transformed poses which have the same origin and scale as reference (now living in world-2 frame)
+        transformed point3s which has minimum error from the reference (now living in world-2 frame).
+        Scale factor of the transformation.
+        Rotation of the transformation.
+        Origin shift (translation) of the transformation.
     """
-    # match the scales first
-    wTi0 = input_list[0]
-    input_distances = np.array([np.linalg.norm((wTi.between(wTi0)).translation()) for wTi in input_list[1:]])
+    # TODO: use GTSAM's align function when it has been merged in develop.
+    N = len(input_list)
+
+    # convert input to 2D arrays, with rows as individual poses
+    input_pose_matrix = np.vstack([x.reshape(1, 3) for x in input_list])
+    ref_pose_matrix = np.vstack([x.reshape(1, 3) for x in ref_list])
+
+    # compute the mean pose
+    input_mean = np.mean(input_pose_matrix, axis=0)
+    ref_mean = np.mean(ref_pose_matrix, axis=0)
+
+    # subtract the means
+    input_pose_matrix = input_pose_matrix - input_mean.reshape(1, 3)
+    ref_pose_matrix = ref_pose_matrix - ref_mean.reshape(1, 3)
+
+    # compute the variances
+    input_variance = np.sum(np.square(input_pose_matrix), axis=None) / N
+    # ref_variance = np.sum(np.square(ref_pose_matrix), axis=None) / N
+
+    # compute the covariance matrix and perform SVD
+    cov_mat = np.dot(ref_pose_matrix.T, input_pose_matrix) / N
+    U, S, Vt = np.linalg.svd(cov_mat)
+    S = np.diag(S)
+
+    W = np.eye(3)
+    if np.linalg.det(U) * np.linalg.det(Vt) < 0:
+        W[2, 2] = -1
+
+    rot_matrix = U @ W @ Vt
+    scale = np.trace(S @ W) / input_variance
+    origin_translation = ref_mean - scale * rot_matrix @ input_mean
+
+    return (
+        [scale * (rot_matrix @ x) + origin_translation for x in input_list],
+        scale,
+        Rot3(rot_matrix),
+        origin_translation,
+    )
 
-    wTi0 = ref_list[0]
-    ref_distances = np.array([np.linalg.norm((wTi.between(wTi0)).translation()) for wTi in ref_list[1:]]) + EPSILON
 
-    # rescale poses to account for SfM scale ambiguity
-    scales = ref_distances / input_distances
-    scaling_factor = np.median(scales)
+def align_poses(input_list: List[Pose3], ref_list: List[Pose3]) -> List[Pose3]:
+    """Aligns the input list of poses to the reference list by aligning rotations followed by translations. The two
+    alignments are done independently.
+
+    Args:
+        input_list: input pose3s which need to be aligned, suppose w1Ti in world-1 frame for all frames i.
+        ref_list: reference pose3s which are target for alignment, suppose w2Ti_ in world-2 frame for all frames i.
```

Comment:
You could also simply use w1Ti_list and w2Ti_list to name the variables. Up to you.

Reply:
I thought it would be nice to not tie it up to a world frame. But after today's meeting, its best to write them as wTa and wTb? Or w1Ta and w2Ta?